File: projects/crates/wordcloud/src/quadtree.rs
//! Quad-tree collision index for layout placement.

/// Axis-aligned box tracked by the quad-tree.
#[derive(Debug, Clone, PartialEq)]
pub struct CollisionBox {
    pub id: String,
    pub x: f32,
    pub y: f32,
    pub width: f32,
    pub height: f32,
}

#[derive(Debug, Clone)]
struct Node {
    x: f32,
    y: f32,
    width: f32,
    height: f32,
    level: u32,
    items: Vec<CollisionBox>,
    children: Option<[Box<Node>; 4]>,
}

/// Spatial index for word placement collision checks.
pub struct QuadTree {
    root: Node,
    max_items: usize,
    max_level: u32,
}

impl QuadTree {
    pub fn new(x: f32, y: f32, width: f32, height: f32) -> Self {
        Self { root: Node::leaf(x, y, width, height, 0), max_items: 10, max_level: 5 }
    }

    pub fn clear(&mut self) {
        let bounds = (self.root.x, self.root.y, self.root.width, self.root.height);
        self.root = Node::leaf(bounds.0, bounds.1, bounds.2, bounds.3, 0);
    }

    pub fn insert(&mut self, item: CollisionBox) {
        Self::insert_into(&mut self.root, item, self.max_items, self.max_level);
    }

    pub fn has_collision(&self, item: &CollisionBox) -> bool {
        let candidates = self.query(item.x, item.y, item.width, item.height);
        candidates.iter().any(|other| other.id != item.id && boxes_overlap(item, other))
    }

    pub fn query(&self, x: f32, y: f32, width: f32, height: f32) -> Vec<CollisionBox> {
        let mut out = Vec::new();
        Self::query_node(&self.root, x, y, width, height, &mut out);
        out
    }

    fn insert_into(node: &mut Node, item: CollisionBox, max_items: usize, max_level: u32) {
        if !rect_intersects(node.x, node.y, node.width, node.height, &item) {
            return;
        }

        if node.children.is_none() && node.items.len() < max_items {
            node.items.push(item);
            return;
        }

        if node.children.is_none() {
            node.subdivide(max_items, max_level);
        }

        if let Some(children) = &mut node.children {
            for child in children.iter_mut() {
                Self::insert_into(child, item.clone(), max_items, max_level);
            }
        }
    }

    fn query_node(node: &Node, x: f32, y: f32, width: f32, height: f32, out: &mut Vec<CollisionBox>) {
        if !regions_overlap(node.x, node.y, node.width, node.height, x, y, width, height) {
            return;
        }

        for item in &node.items {
            if regions_overlap(item.x, item.y, item.width, item.height, x, y, width, height) {
                out.push(item.clone());
            }
        }

        if let Some(children) = &node.children {
            for child in children.iter() {
                Self::query_node(child, x, y, width, height, out);
            }
        }
    }
}

impl Node {
    fn leaf(x: f32, y: f32, width: f32, height: f32, level: u32) -> Self {
        Self { x, y, width, height, level, items: Vec::new(), children: None }
    }

    fn subdivide(&mut self, max_items: usize, max_level: u32) {
        if self.level >= max_level {
            return;
        }

        let existing = self.items.clone();
        let half_w = self.width / 2.0;
        let half_h = self.height / 2.0;
        let level = self.level + 1;

        self.children = Some([
            Box::new(Node::leaf(self.x, self.y, half_w, half_h, level)),
            Box::new(Node::leaf(self.x + half_w, self.y, half_w, half_h, level)),
            Box::new(Node::leaf(self.x, self.y + half_h, half_w, half_h, level)),
            Box::new(Node::leaf(self.x + half_w, self.y + half_h, half_w, half_h, level)),
        ]);
        self.items.clear();

        if let Some(children) = &mut self.children {
            for item in existing {
                for child in children.iter_mut() {
                    QuadTree::insert_into(child, item.clone(), max_items, max_level);
                }
            }
        }
    }
}

fn boxes_overlap(a: &CollisionBox, b: &CollisionBox) -> bool {
    regions_overlap(a.x, a.y, a.width, a.height, b.x, b.y, b.width, b.height)
}

fn rect_intersects(nx: f32, ny: f32, nw: f32, nh: f32, item: &CollisionBox) -> bool {
    regions_overlap(nx, ny, nw, nh, item.x, item.y, item.width, item.height)
}

fn regions_overlap(ax: f32, ay: f32, aw: f32, ah: f32, bx: f32, by: f32, bw: f32, bh: f32) -> bool {
    !(ax + aw < bx || bx + bw < ax || ay + ah < by || by + bh < ay)
}

File: projects/wordcloud-ts/src/quadtree.ts
import { QuadTreeNode, WordCloudElement } from './types';

/**
 * 四叉树实现，用于高效的碰撞检测和空间管理
 */
export class QuadTree {
  private root: QuadTreeNode;
  private readonly maxElements: number;
  private readonly maxLevel: number;

  constructor(
    x: number,
    y: number,
    width: number,
    height: number,
    maxElements: number = 10,
    maxLevel: number = 5
  ) {
    this.maxElements = maxElements;
    this.maxLevel = maxLevel;
    this.root = {
      x,
      y,
      width,
      height,
      elements: [],
      maxElements,
      level: 0
    };
  }

  /**
   * 插入元素到四叉树
   */
  insert(element: WordCloudElement): void {
    this.insertIntoNode(this.root, element);
  }

  private insertIntoNode(node: QuadTreeNode, element: WordCloudElement): void {
    // 检查元素是否在节点范围内
    if (!this.intersects(node, element)) {
      return;
    }

    // 如果节点未分割且元素数量未超限
    if (!node.children && node.elements.length < node.maxElements) {
      node.elements.push(element);
      return;
    }

    // 如果节点未分割但需要分割
    if (!node.children) {
      this.subdivide(node);
    }

    // 尝试插入到子节点
    if (node.children) {
      for (const child of node.children) {
        this.insertIntoNode(child, element);
      }
    }
  }

  /**
   * 分割节点为四个子节点
   */
  private subdivide(node: QuadTreeNode): void {
    if (node.level >= this.maxLevel) {
      return;
    }

    const halfWidth = node.width / 2;
    const halfHeight = node.height / 2;
    const level = node.level + 1;

    node.children = [
      // 左上
      {
        x: node.x,
        y: node.y,
        width: halfWidth,
        height: halfHeight,
        elements: [],
        maxElements: this.maxElements,
        level
      },
      // 右上
      {
        x: node.x + halfWidth,
        y: node.y,
        width: halfWidth,
        height: halfHeight,
        elements: [],
        maxElements: this.maxElements,
        level
      },
      // 左下
      {
        x: node.x,
        y: node.y + halfHeight,
        width: halfWidth,
        height: halfHeight,
        elements: [],
        maxElements: this.maxElements,
        level
      },
      // 右下
      {
        x: node.x + halfWidth,
        y: node.y + halfHeight,
        width: halfWidth,
        height: halfHeight,
        elements: [],
        maxElements: this.maxElements,
        level
      }
    ];

    // 重新分配现有元素到子节点
    for (const element of node.elements) {
      for (const child of node.children) {
        this.insertIntoNode(child, element);
      }
    }
    node.elements = [];
  }

  /**
   * 检查元素是否与节点相交
   */
  private intersects(node: QuadTreeNode, element: WordCloudElement): boolean {
    return !(
      element.x > node.x + node.width ||
      element.x + element.width < node.x ||
      element.y > node.y + node.height ||
      element.y + element.height < node.y
    );
  }

  /**
   * 查询与给定区域相交的所有元素
   */
  query(x: number, y: number, width: number, height: number): WordCloudElement[] {
    const result: WordCloudElement[] = [];
    this.queryNode(this.root, x, y, width, height, result);
    return result;
  }

  private queryNode(
    node: QuadTreeNode,
    x: number,
    y: number,
    width: number,
    height: number,
    result: WordCloudElement[]
  ): void {
    // 检查查询区域是否与节点相交
    if (
      x > node.x + node.width ||
      x + width < node.x ||
      y > node.y + node.height ||
      y + height < node.y
    ) {
      return;
    }

    // 添加节点中的元素
    for (const element of node.elements) {
      if (this.elementIntersects(element, x, y, width, height)) {
        result.push(element);
      }
    }

    // 递归查询子节点
    if (node.children) {
      for (const child of node.children) {
        this.queryNode(child, x, y, width, height, result);
      }
    }
  }

  /**
   * 检查元素是否与指定区域相交
   */
  private elementIntersects(
    element: WordCloudElement,
    x: number,
    y: number,
    width: number,
    height: number
  ): boolean {
    return !(
      element.x > x + width ||
      element.x + element.width < x ||
      element.y > y + height ||
      element.y + element.height < y
    );
  }

  /**
   * 检查新元素是否与现有元素碰撞
   */
  hasCollision(element: WordCloudElement): boolean {
    const candidates = this.query(
      element.x,
      element.y,
      element.width,
      element.height
    );

    return candidates.some(candidate => 
      candidate.id !== element.id && this.elementsCollide(element, candidate)
    );
  }

  /**
   * 检查两个元素是否碰撞
   */
  private elementsCollide(a: WordCloudElement, b: WordCloudElement): boolean {
    return !(
      a.x > b.x + b.width ||
      a.x + a.width < b.x ||
      a.y > b.y + b.height ||
      a.y + a.height < b.y
    );
  }

  /**
   * 清空四叉树
   */
  clear(): void {
    this.root.elements = [];
    this.root.children = undefined;
  }

  /**
   * 获取所有元素
   */
  getAllElements(): WordCloudElement[] {
    const result: WordCloudElement[] = [];
    this.collectElements(this.root, result);
    return result;
  }

  private collectElements(node: QuadTreeNode, result: WordCloudElement[]): void {
    result.push(...node.elements);
    if (node.children) {
      for (const child of node.children) {
        this.collectElements(child, result);
      }
    }
  }
}